File: 66-Plus One/plus_one.go
package main

import "fmt"

func main() {
	nums := []int{9}
	fmt.Println("Result: ", plusOne(nums))
}

func plusOne(digits []int) []int {
	//
	// First attempt Jun 02 2024 12:54
	// Runtime: 1ms, Memory: 2.28MB
	//
	for i := len(digits) - 1; i >= 0; i-- {
		if digits[i] != 9 {
			digits[i]++
			return digits
		}

		digits[i] = 0
	}

	return append([]int{1}, digits...)
}

File: 66-Plus One/plus_one.dart
main() {
  final nums = [4, 3, 2, 1];

  final result = PlusOne.plusOne(nums);
  print(result);
}

abstract class PlusOne {
  static List<int> plusOne(List<int> digits) {
    //
    // First attempt Jun 02 2024 13:53
    // Runtime: 340ms, Memory: 148.24MB
    //
    for (int i = digits.length - 1; i >= 0; i--) {
      if (digits[i] != 9) {
        digits[i] = digits[i] + 1;
        return digits;
      }

      digits[i] = 0;
    }

    digits.insert(0, 1);
    return digits;
  }
}
